client/network/src/gossip.rs

```diff
@@ -0,0 +1,318 @@
+	/// Use this method only if the message is extremely important.
+	pub fn push_unbounded(&mut self, message: M) {
+		self.lock.messages_queue.push_back(message);
+	}
+
+	/// Calls `filter` for each message in the queue, and removes the ones for which `false` is
+	/// returned.
+	///
+	/// > **Note**: The parameter of `filter` is a `&M` and not a `&mut M` (which would be
+	/// >           better) because the underlying implementation relies on `VecDeque::retain`.
+	pub fn retain(&mut self, filter: impl FnMut(&M) -> bool) {
+		self.lock.messages_queue.retain(filter);
+	}
+}
+
+impl<'a, M> Drop for QueueLock<'a, M> {
+	fn drop(&mut self) {
+		// We notify the `Condvar` in the destructor in order to be able to push multiple
+		// messages and wake up the background task only one afterwards.
+		self.condvar.notify_one();
+	}
+}
+
+#[derive(Debug)]
+struct Shared<M> {
+	/// Read by the background task after locking `locked`. If true, the task stops.
+	stop_task: atomic::AtomicBool,
+	locked: Mutex<SharedLock<M>>,
+	/// Must be notified every time the content of `locked` changes.
+	condvar: Condvar,
+	/// Maximum number of elements in `messages_queue`.
+	queue_size_limit: usize,
+}
+
+#[derive(Debug)]
+struct SharedLock<M> {
+	/// Queue of messages waiting to be sent out.
+	messages_queue: VecDeque<M>,
+}
+
+async fn spawn_task<M, F: Fn(M) -> Vec<u8>>(
+	service: Arc<dyn AbstractNotificationSender + Send + Sync + 'static>,
+	peer_id: PeerId,
+	protocol: ConsensusEngineId,
+	shared: Arc<Shared<M>>,
+	messages_encode: F,
+) {
+	loop {
+		let next_message = 'next_msg: loop {
+			let mut locked = shared.locked.lock().await;
+
+			loop {
+				if shared.stop_task.load(atomic::Ordering::Acquire) {
+					return;
+				}
+
+				if let Some(msg) = locked.messages_queue.pop_front() {
+					break 'next_msg msg;
+				}
+
+				// It is possible that the destructor of `DirectedGossip` sets `stop_task` to
+				// true and notifies the `Condvar` after the background task loads `stop_task`
+				// and before it calls `Condvar::wait`.
+				// See also the corresponding comment in `DirectedGossip::drop`.
+				// For this reason, we use `wait_timeout`. In the worst case scenario,
+				// `stop_task` will always be checked again after the timeout is reached.
+				locked = shared.condvar.wait_timeout(locked, Duration::from_secs(10)).await.0;
+			}
+		};
+
+		// Starting from below, we try to send the message. If an error happens when sending,
+		// the only sane option we have is to silently discard the message.
+		let sender = match service.notification_sender(peer_id.clone(), protocol) {
+			Ok(s) => s,
+			Err(_) => continue,
+		};
+
+		let ready = match sender.ready().await {
+			Ok(r) => r,
+			Err(_) => continue,
+		};
+
+		let _ = ready.send(messages_encode(next_message));
+	}
+}
+
+/// Abstraction around `NetworkService` that permits removing the `B` and `H` parameters.
+trait AbstractNotificationSender {
```

Comment:
I suggest to remove this trait. The two type parameters `B` and `H` are only necessary in the private `spawn_task` method and adding those is far less code than this trait plus impl. And your calls to `NetworkService::notification_sender` are not dynamically dispatched.

Reply:
It's necessary for `DirectedGossipPrototype` to work. See your other comment.